### ISSUE TITLE
Prevent adding duplicates to playlists.

### ISSUE THREAD

**Issue by unoukujou** [futo-org/grayjay-android#1244]
### What happened?

When adding a video to a playlist, show some indication if the video already exists in that playlist to prevent the user from adding the same video multiple times.

Happened to me already a couple of times now and I have to go through my playlist to check for duplicates.

### Grayjay Version

all

### What plugins are you seeing the problem on?

All

### When do you experience the issue?

- [ ] While logged in
- [ ] While logged out
- [ ] N/A

**Comment by PhilC813** [comment]
In the slide-in menu where you can add the video to playlists, there should also be some sort of indicator to specify which playlists already contain said video.

**Comment by unoukujou** [comment]
> In the slide-in menu where you can add the video to playlists, there should also be some sort of indicator to specify which playlists already contain said video.

Not seeing it. Can you share a screenshot?

**Comment by PhilC813** [comment]
> Nevermind, I think you're just suggesting that it should be that way?

Yup. So the UI should make it clear that a video is already present in a playlist. Then, as you suggest, have detection detection and prevention.

**Comment by TheNewsYorkTime** [comment]
I think that duplicates should not be possible. 
Also, somewhat related, I think here should be status "popups" for more actions. Specifically playlist actions. Like at the bottom of the screen a non-interactable non-blocking element that says whether certain actions are successful or not. they already have it, like for informing you that a channel was removed from the source or that there's no internet / failed to doenload or open a video. It could probably be used in this case. "Add to playlist failed - duplicate"

**Comment by PhilC813** [comment]
> Also, somewhat related, I think here should be status "popups" for more actions. Specifically playlist actions. Like at the bottom of the screen a non-interactable non-blocking element that says whether certain actions are successful or not. they already have it, like for informing you that a channel was removed from the source or that there's no internet / failed to doenload or open a video.

I guess you're referring to this kind of UI element:
https://developer.android.com/guide/topics/ui/notifiers/toasts

> It could probably be used in this case. "Add to playlist failed - duplicate"

Another way would be to prevent the user of selecting the playlist in the first place, like make it greyed out in the previous menu. But otherwise yes, that would definitely be useful; the app is not "verbose" enough outside of the video player.

**Comment by TheNewsYorkTime** [comment]
Yes, exactly.

**Comment by TheNewsYorkTime** [comment]
Reading about toasts, snackbars would propably be more appropriate in this scenario.
[https://m3.material.io/components/snackbar/overview](https://m3.material.io/components/snackbar/overview)
There are tradeoffs for both, depending on where they are located. If a snackbar covers the bottom navigation buttons then that would be bad. There are also times when toasts would be better, like some apps (I forget which) have an "entering fullscreen" toast. 

I do think that just making it impossible to add a duplicate to a playlist is not enough. It puts a lot of pressure on the developer to remove every way of adding duplicates to a playlist. And if one got in, then it would still need removed. I think duplicates should be reomoved "on sight".  

The thing is, some people want duplicates, ie comments in  #305. Some people want to have a video repeated multiple times for whatever reason. They want to sleep, or they really like a song. So there should be a setting to enable duplicates. That setting would enable a option to remove duplicates from a playlist after selecting it, as well as removing your idea of restriction on adding duplicate videos to playlists. And some playlists could be set as duplicate free, to re-enable the block for them. Or the setting could be less influential and there could be a duplicate enabled playlist button, instead of most of what I just said. Obviously the removal of duplicates automatically would not happen for duplicate enabled playlists. 

Also, in whatever world, I think that the queue should be duplicate enabled by default. 

Also I think that *part* of the duplicate issue would be solved with more verbosity, so you know the add succeeded, and don't add a video multiple times in a row.

**Comment by Zvonimir-FUTO** [comment]
This has been implemented.
